### PR TITLE
Api changes

### PR DESCRIPTION
**Pull request by jlmucb**
Squashed merge message

API changes intended to make high level interface transparent, robust and concise in the face of future development.

New initialize_enclave, simplified individualized inits, simplified server_dispatch and init_client/server_ssl.
cc_trust_data is now cc_trust_manager.  Fixed bugs in application_enclave

### CODE REVIEW

**Review by gapisback — CHANGES_REQUESTED** [review]
John, let's comment out pytests code first to get a look into stability of the rest of the tests.

I will help you overhaul SWIG interfaces & pytests in a few days.

**Comment by gapisback** [comment]
@jlmucb - Somehow my code-level comment seems to have been lost.

I think your interface changes are not reflected in pytests / SWIG code. This is why [your CI-job ](https://github.com/vmware-research/certifier-framework-for-confidential-computing/actions/runs/6078151663/job/16488942658?pr=190)is now bombing completely early-on for pytests.

Re-jiggering pytests & SWIG to keep up with interfaces is a continual effort.

I suggest here's what you should do:

- Delete this whole test code block named `name: test-cert_framework-pytests` from `build.yml`.
- Check that into your dev branch, which will trigger CI. You'll get to see how the rest of the tests perform.
- Let's try to get the rest of the review and your changes finalized thru review this week.

Later this week, I will find time to rework SWIG interfaces & pytests code to be in-sync with your newer interfaces. I will provide a delta-commit on your dev-branch just to re-engage and re-stabilize pytests.

Hope this works for you & @yelvmw

File: Doc/developer_tour.md [hunk 1]
```diff
@@ -65,35 +65,39 @@ Here is an example of such a program and its interactions with the Certifier API
...
 
 
   // Open secure channel
   string my_role("client");  // or "server"
   secure_authenticated_channel channel(my_role);
   channel.init_client_ssl(FLAGS_server_app_host, FLAGS_server_app_port,
-          app_trust_data->serialized_policy_cert_, app_trust_data->private_auth_key_,
-          app_trust_data->serialized_primary_admissions_cert_);
+          trust_mgr->serialized_policy_cert_, trust_mgr->private_auth_key_,
+          trust_mgr->serialized_primary_admissions_cert_);
```
Comment (yelvmw) [inline Doc/developer_tour.md:100]: So we are not updating the init_client_ssl interface to take a "trust_mgr" object? If this is the case, we need to remove the protobuf type from the parameter and take key properties and buffers instead.

File: include/certifier_framework.h [hunk 2]
```diff
@@ -377,6 +392,12 @@ bool server_dispatch(const string &host_name,
                      key_message & private_key,
                      const string &private_key_cert,
                      void (*)(secure_authenticated_channel &));
+
+bool server_dispatch(const string &          host_name,
```
Comment (yelvmw) [inline include/certifier_framework.h:396]: Have we considered moving this into secure_authenticated_channel class?

File: include/certifier_framework.h [hunk 3]
```diff
@@ -361,6 +369,13 @@ class secure_authenticated_channel {
...
 
+  bool init_client_ssl(const string &          host_name,
+                       int                     port,
+                       const cc_trust_manager &mgr);
+  bool init_server_ssl(const string &          host_name,
+                       int                     port,
+                       const cc_trust_manager &mgr);
+
   void server_channel_accept_and_auth(
       void (*func)(secure_authenticated_channel &));
```
Comment (yelvmw) [inline include/certifier_framework.h:381]: Didn't we agree to refactor the read/write interfaces?

Comment (jlmucb) [reply include/certifier_framework.h:381]: I did except for readmsg and writemsg where I think there are problems.  I'd do that in a subsequent pull if we decide to proceed.

**Review by yelvmw — APPROVED** [review]
I think this is good for upstream. We still need to do the following but they can be deferred to later patches:

1. Update simple apps to use the simplified interfaces. We could do this when we start refactoring simple apps.
2. Discuss and follow-up on secure_authenticated_channel read/write interfaces.

Yes

File: include/certifier_utilities.h [hunk 4]
```diff
@@ -34,10 +34,13 @@ namespace utilities {
 const int block_size = 16;
 const int num_bits_in_byte = 8;
 
-bool write_file(const string &file_name, int size, byte *data);
-int  file_size(const string &file_name);
+int file_size(const string &file_name);
+
 bool read_file(const string &file_name, int *size, byte *data);
+bool write_file(const string &file_name, int size, byte *data);
```
Comment (gapisback) [inline include/certifier_utilities.h:40]: In the proposal doc we reviewed, there is a mention of reworking read() / write() interfaces.

I see these new ones being added to this file, but the ones in `certifier_framework.h` still seem to remain unchanged; see this triplet below:

```
382   int  read(string *out);
383   int  read(int size, byte *b);
384   int  write(int size, byte *b);
```
Qs. is: Are these new interfaces a substitute for the existing ones? And, there is a mention in the doc to add some extended interfaces. Is there a plan to go back nand address the issues w.r.t above 3 interfaces which we discussed earlier?

**Review by gapisback — CHANGES_REQUESTED** [review]
Hi, @jlmucb - I completed one round of review for the core interface .h / .c files.

I will try to finish the rest later today. Some items may merit revision in this round, hence, I have marked this as "Request changes".

File: include/certifier_framework.h [hunk 5]
```diff
@@ -221,34 +221,42 @@ class cc_trust_data {
...
-  ~cc_trust_data();
+  cc_trust_manager();
+  cc_trust_manager(const string &enclave_type,
+                   const string &purpose,
+                   const string &policy_store_name);
+  ~cc_trust_manager();
+
+  // If n == 0, systems should be able to find parameters
+  // by default, for example, sev and sgx.
+  bool initialize_enclave(int n, string *params);
```
[the remark this answers is not in the document]
Comment (jlmucb) [reply include/certifier_framework.h:232]: This is inconsistant with existing code so I'll defer this.

File: sample_apps/multidomain_simple_app/multidomain_client_app.cc [hunk 6]
```diff
@@ -108,23 +108,31 @@ bool get_simulated_enclave_parameters(string **s, int *n) {
...
+           __func__,
+           __LINE__,
+           err.c_str());
     return false;
   }
 
   if (!read_file_into_string(FLAGS_data_dir + FLAGS_measurement_file,
                              &args[1])) {
-    printf("%s() error, line %d, Can't read measurement file\n",
+    string err = FLAGS_data_dir + FLAGS_measurement_file;
```
Comment (gapisback) [inline sample_apps/multidomain_simple_app/multidomain_client_app.cc:121]: I think this redeclaration of `string err` should be unnecessary. You should be able to use the `err` variable declared on L111.

File: sample_apps/multidomain_simple_app/multidomain_client_app.cc [hunk 7]
```diff
@@ -108,23 +108,31 @@ bool get_simulated_enclave_parameters(string **s, int *n) {
...
+    printf("%s() error, line %d, Can't read measurement file %s\n",
            __func__,
-           __LINE__);
+           __LINE__,
+           err.c_str().);
```
Comment (gapisback) [inline sample_apps/multidomain_simple_app/multidomain_client_app.cc:125]: How is this compiling ...? `err.c_str().);` ....?

Delete the trailing `.`

File: sample_apps/multidomain_simple_app/multidomain_client_app.cc [hunk 8]
```diff
@@ -108,23 +108,31 @@ bool get_simulated_enclave_parameters(string **s, int *n) {
...
     return false;
   }
 
   if (!read_file_into_string(FLAGS_data_dir + FLAGS_platform_attest_endorsement,
                              &args[2])) {
-    printf("%s() error, line %d, Can't read endorsement file\n",
+    string err = FLAGS_data_dir + FLAGS_platform_attest_endorsement;
```
Comment (gapisback) [inline sample_apps/multidomain_simple_app/multidomain_client_app.cc:131]: Remove `string`.

**Review by gapisback — APPROVED** [review]
Took a quick re-look. Some minor cleanup is pending ...

File: sample_apps/multidomain_simple_app/multidomain_client_app.cc (hunk 8 above)
Comment (jlmucb) [reply sample_apps/multidomain_simple_app/multidomain_client_app.cc:131]: ?.  This is used to print the file name you asked for.

Comment (gapisback) [reply sample_apps/multidomain_simple_app/multidomain_client_app.cc:131]: I'm suggesting the same thing I suggested for L121; i.e. you should be able to simply say:

`err = FLAGS_data_dir + FLAGS_platform_attest_endorsement;`

and not have to declare `string err = ...` again.

The previous declaration of `err` should work, I think.

File: sample_apps/multidomain_simple_app/multidomain_client_app.cc (hunk 6 above)
Comment (jlmucb) [reply sample_apps/multidomain_simple_app/multidomain_client_app.cc:121]: I'd rather keep this local to the block.

File: sample_apps/multidomain_simple_app/multidomain_client_app.cc (hunk 7 above)
Comment (jlmucb) [reply sample_apps/multidomain_simple_app/multidomain_client_app.cc]: fixed

**Pull request merged**